### PR TITLE
PCHR-3621: Fix issue with Manager Not Being Able to Approve Leave Request on Period End Date

### PR DESCRIPTION
**Pull request by tunbola**
## Overview
When a staff requests for leave on the end date of a period, when the leave manager attempts to approve such leave request, even though the staff has entitlement, an error is displayed and the manager is unable to approve the leave request.

## Before
- When staff requests leave with date falling on the end date of the absence period, the manager is unable to approve such request.

beforeenddate

## After
- When staff requests leave with date falling on the end date of the absence period, the manager is able to approve such request.

afterenddate

## Technical Details
The function responsible for finding the absence period containing the Leave request dates was comparing the dates in different formats. The leave dates are in date time format while the period dates are in date format.
E.g
Absence period- start Date: 2018-01-01
Absence Period- end Sate: 2018-12-31

Leave Date- from Date: 2018-12-31 00:00:00
Leave Date- to Date: 2018-12-31 23:59:00

The function was changed to compare the dates in same format because the original function implementation always had the leave end date greater than the period end date

```php
    function isRequestInPeriod (request, period) {
      var requestFromDate = moment(request.from_date, 'YYYY-MM-DD');
      var requestToDate = moment(request.to_date, 'YYYY-MM-DD');

      return (requestFromDate.isSameOrAfter(period.start_date) &&
        requestToDate.isSameOrBefore(period.end_date));
    }
```

### CODE REVIEW

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js [hunk 1]
```diff
@@ -664,8 +664,8 @@ define([
      * @return {Boolean}
      */
     function isRequestInPeriod (request, period) {
-      var requestFromDate = moment(request.from_date);
-      var requestToDate = moment(request.to_date);
+      var requestFromDate = moment(request.from_date, 'YYYY-MM-DD');
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js]: You don't really need to convert explicitly, you may try to compare later like that 

```js
var requestFromDate = moment(request.from_date);
// ...
// just tell it to compare dates (it also considers month and year)
return (requestFromDate.isSameOrAfter(period.start_date, 'day')...
```

Please have a look at https://momentjs.com/docs/#/query/is-same-or-after/ for reference.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/request.controller.spec.js [hunk 2]
```diff
@@ -593,261 +593,288 @@
       });
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/request.controller.spec.js]: Please revert changes and make two separate commits:
1) only wrap into a separate describe, nothing else
2) include new tests

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/request.controller.spec.js [hunk 3]
```diff
@@ -593,6 +593,31 @@
...
 
       describe('manager opens leave request popup in edit mode', function () {
+        describe('when leave request date has same date as period end date', function () {
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/request.controller.spec.js]: `when leave request "to" date is same as absence period end date`

We also miss a case

`when leave request "from" date is same as absence period start date`

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/request.controller.spec.js [hunk 4]
```diff
@@ -593,6 +593,31 @@
...
+          beforeEach(function () {
+            var approvalStatus = optionGroupMock.specificValue('hrleaveandabsences_leave_request_status', 'value', '1');
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/request.controller.spec.js]: Because we now check if absence period was defined by leave request dates (and not actually checking if it was submitted successfully), we may not need to actually submit it. So this line may go.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/request.controller.spec.js [hunk 5]
```diff
@@ -593,6 +593,31 @@
...
+            leaveRequest.from_date = '2017-12-31 00:00:00';
+            leaveRequest.to_date = '2017-12-31 23:59:00';
+            leaveRequest.contact_id = CRM.vars.leaveAndAbsences.contactId.toString();
+            role = 'manager';
+
+            initTestController({ leaveRequest: leaveRequest });
+            spyOn(controller.request, 'update').and.callThrough();
+            spyOn(LeaveRequestInstance, 'calculateBalanceChange').and.returnValue(
+              $q.resolve({ amount: controller.balance.change.amount }));
+            leaveRequest.status_id = approvalStatus;
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/request.controller.spec.js]: Same as above, may not be needed.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/request.controller.spec.js [hunk 6]
```diff
@@ -593,6 +593,31 @@
...
+            controller.submit();
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/request.controller.spec.js]: Same as above, may not be needed.

**Comment by tunbola** [comment]
@igorpavlov , The commit messages were bearing the wrong ticket number 3261 instead of 3621 so I fixed that.
